### PR TITLE
Make Clio use async database reads

### LINKED ISSUE #44: Make database fetches truly async
Currently, we use cassandra's async API to fetch records, but we really just submit a batch of async requests and then wait on a condition variable until they all complete. It would be better to actually free up the thread that makes the async requests, so that way the thread can handle additional RPC calls while waiting for the async requests to complete. This will give us better throughput overall for reads.

We need to figure out how to make async postgres queries.

It would be best to change all database reads to be asynchronous, even the ones that currently use the synchronous cassandra API.

### PR DESCRIPTION
**Pull request by natenichols**
Uses `Boost.Asio` to make database reads (and writes in postgres) async. Forwarding also uses async reads/writes

fixes #44 

Lets start w/ some high level stuff.

Whenever a websocket or http server is handling a request, instead of handling that request synchronously, Clio will now spawn a co-routine with `boost::asio::spawn` to handle the request. This provides us with the yield context we will use to yield the coroutine when awaiting database responses.

**Cassandra:**
In cassandra, I'm only changing the database reads, the writes are already async, so we'll just leave those as is. This is pretty simple, all `CassandraBackend` methods use three methods to use do database reads `executeSyncRead`, `fetchTransactions`, and `fetchLedgerObjects`. These are the three we'll have to make async.

For `executeSyncRead`, its pretty straightforward, we rewrite `cass_future_error_code` using an async_result, which waits asynchronously. 

For `fetchTransactions` and `fetchLedgerObjects`, we just replace the `condition_variable` with an `async_result::completion_handler_type` to be called when `outstanding == 0`.

**Postgres:**
Postgres is a lot harder, since we use `PgPool` to manage connections. This leads to a lot of open connections in postgres (which is fine, postgres can handle it). I rewrote the `Pg` class to be completely async, which means the number of `max_connections` in the PgPool must be set to be high. 

We don't have to use `async_result` here, since libPQ can expose the socket underlying each postgres connection, so we can async_wait on the socket for either `read-ready` or `write-ready` depending on the status of the connection.

Because the PgClass is now async, this means both reads and writes are now async, so I had to spawn coroutines to handle the ETL processes.

Note: I also added a `PgQuerySync` class, that does sync queries, for setting up the databases when the `io_context` is not running. This should not be used outside of setting up the database.

---
This change is [Reviewable](https://reviewable.io/reviews/cjcobb23/clio/76)

### CODE REVIEW

File: src/backend/Pg.cpp
Comment (mtrippled) [inline src/backend/Pg.cpp:262]: I'm pretty sure PQsendQueryParams() with 0 nParams and null values is the same as PQsendQuery(). It's handled the same as PQexec & PQexecParams (the synchronous version): https://www.postgresql.org/docs/12/libpq-exec.html

So there should be no reason for calling the separate functions, nor for keeping the sent variable since the status can be checked for a single function call.

File: src/backend/Pg.cpp
Comment (mtrippled) [inline src/backend/Pg.cpp:314]: What do you want to do about multiple results? One possibility is to check if there's already a result, and then throw:

            if (ret)
            {
                Throw<std::runtime_error>("multiple results returned");
            }

The way it stands now is that if there are multiple commands submitted in the command*, such as "SELECT * FROM foo; SELECT * FROM bar", then this will return the results from the 2nd query.

Not a big deal either way. One way forces a single SQL statement per command submissions, the other allows multiple but silently loses the response to all but the last.

File: src/backend/Pg.cpp
Comment (mtrippled) [inline src/backend/Pg.cpp:575]: I'm actually looking at the code already implemented, and it looks like it forces the user to connect with username, password, and database name. In reporting mode, if the user does not specify these things, the connection defaults to using postgres on the local system as the same username/dbname as the rippled process.

Just a suggestion to make things easier for users running this on a single server.

Upon further thought, I'm not sure if asio would work with postgres on a local UNIX domain socket. So this requires the DB to be accessed over the network (even localhost), which requires specific configuration.

File: src/backend/Pg.h
Comment (mtrippled) [inline src/backend/Pg.h:452]: The idleSweeper() is not called by anything.

File: unittests/main.cpp
Comment (mtrippled) [inline unittests/main.cpp:2133]: Not in this PR, but in the code anyway, when doing checkin(), in addition to clearing the results, the asio socket should be cancelled. This aborts any pending asio async operations.

    /** Cancel asynchronous operations on underlying postgres socket. */
    void
    cancel()
    {
        if (socket_)
            try { socket_->cancel(); } catch (...) {}
    }

**Review by mtrippled — CHANGES_REQUESTED** [review]
This is a first pass at the PR. I looked only at Pg.h and Pg.cpp, and also included some comments about the pre-existing code there.

File: src/backend/Pg.cpp
Comment (natenichols) [reply src/backend/Pg.cpp:314]: So I think we want to allow multiple queries for setting up the database and creating tables, so we do want to allow this behavior.

Comment (natenichols) [reply src/backend/Pg.cpp:314]: In that case, if one errors out, we should break and return the error. Otherwise a final success message should imply that all the queries were successful.

Comment (mtrippled) [reply src/backend/Pg.cpp:314]: That's fine.

File: src/backend/Pg.h
Comment (cjcobb23) [reply src/backend/Pg.h:261]: Is there a problem if a different thread completes the async activity? i.e. Thread A starts, Thread B finishes

File: src/backend/Pg.cpp
Comment (cjcobb23) [reply src/backend/Pg.cpp:458]: There are times when we don't want copy to be synchronous. There are multiple tables that all need to be copied to, but if copy is synchronous, we need to wait for one to finish before submitting the next. Also, during the initial ledger load, we flush the buffer via copy every 100,000 or million records or so.

File: src/etl/ETLSource.h
Comment (cjcobb23) [inline src/etl/ETLSource.h:57]: Can we make the yield_context a member variable?

File: src/backend/Pg.h
Comment (natenichols) [reply src/backend/Pg.h:265]: So I actually need to refactor this. According to the docs 
```
Obtains the file descriptor number of the connection socket to the server. A valid descriptor will be greater than or equal to 0; a result of -1 indicates that no server connection is currently open. (This will not change during normal operation, but could change during connection setup or reset.)
```
So it looks like we only need to acquire the socket during connection setup or reset.

Comment (mtrippled) [reply src/backend/Pg.h:265]: Yes, same with connect()

File: src/backend/Pg.cpp
Comment (mtrippled) [reply src/backend/Pg.cpp:458]: Are they all in the same transaction, such as in reporting mode?

Anyway, see the comment above this for Pg.cpp:454. Namely, this function looks broken to me, or at least not working as postgres documents it. Fixing it by making it synchronous would be easier.

File: src/backend/Pg.cpp
Comment (natenichols) [reply src/backend/Pg.cpp:470]: Cool, I'll wrap all async_waits w/ try/catch blocks.

File: src/backend/Pg.cpp
Comment (cjcobb23) [reply src/backend/Pg.cpp:458]: yeah they are all in the same transaction

File: src/backend/Pg.h
Comment (natenichols) [reply src/backend/Pg.h:265]: So I have made the change to 
1) only acquire socket during connection, since it will not change during normal operation.
According to the docs:
"If PQconnectStart succeeds, the next stage is to poll libpq so that it can proceed with the connection sequence. Use PQsocket(conn) to obtain the descriptor of the socket underlying the database connection"
So as long as PQConnectStart succeeds we should be good to obtain the underlying socket.

2) Use RAII to manage the socket, calling `socket_->cancel()` upon destruction.

File: src/backend/Pg.cpp
Comment (natenichols) [reply src/backend/Pg.cpp:458]: My understanding is that synchronous `COPY` should work similarly to `query`, but calling `PQputCopyData` or `PQputCopyEnd` instead of `PQsendQueryParams`.

This does _work_, I'm gonna look through the documentation to try to find how to correctly implement it.

Comment (natenichols) [reply src/backend/Pg.cpp]: So after looking at some docs, this looks correct 
" To prevent this possibility and achieve completely nonblocking database operation, the following additional functions can be used.", and then lists `PQFlush`.

We flush after we send each command. 
"After sending any command or data on a nonblocking connection, call PQflush. If it returns 1, wait for the socket to become read- or write-ready. If it becomes write-ready, call PQflush again. If it becomes read-ready, call PQconsumeInput , then call PQflush again. Repeat until PQflush returns 0. (It is necessary to check for read-ready and drain the input with PQconsumeInput , because the server can block trying to send us data, e.g., NOTICE messages, and won't read our data until we read its.) Once PQflush returns 0, wait for the socket to be read-ready and then read the response as described above." 

Using PQflush here should eliminate the issue described above, if I'm RTFM-ing correctly.

File: src/backend/Pg.cpp
Comment (mtrippled) [reply src/backend/Pg.cpp]: Duh, I'm almost as bad at reading code as I am at writing it.

File: src/backend/BackendInterface.h
Comment (mtrippled) [reply src/backend/BackendInterface.h:24]: The problem then is manually synchronizing. strands fix that, but by necessity they require the same io_context. Just have 1 big threadpool and tune it. It's not like if cassandra or postgres timeout all over the place everything else works. If they break, it's all broken.

File: src/backend/Pg.h
Comment (natenichols) [reply src/backend/Pg.h:262]: No, there shouldn't be a problem if Thread A suspends a coro, and Thread B resumes it, as long as Thread A does not hold  any locks.

File: src/backend/CassandraBackend.h
Comment (natenichols) [reply src/backend/CassandraBackend.h:684]: Yeah, the postgres impl needs it

File: src/backend/CassandraBackend.h
Comment (natenichols) [reply src/backend/CassandraBackend.h:890]: Same

File: src/backend/PostgresBackend.cpp
Comment (undertome) [inline src/backend/PostgresBackend.cpp:459]: If we're anticipating that `results` will contain roughly `hashes.size()` entries, consider increasing the vector's capacity with `reserve`.

File: src/backend/PostgresBackend.cpp
Comment (undertome) [inline src/backend/PostgresBackend.cpp:502]: If we're anticipating that `results` will contain roughly `keys.size()` entries, consider increasing the vector's capacity with `reserve`.

File: src/backend/PostgresBackend.cpp
Comment (undertome) [inline src/backend/PostgresBackend.cpp:473]: This may exceed the scope of this pull request, and I'm fine with that. However, there are several places in the code where we time the execution of various functions. Consider using [`boost::cpu_timer`](https://www.boost.org/doc/libs/1_75_0/libs/timer/doc/cpu_timers.html) to avoid manually calculating the duration.

File: src/backend/PostgresBackend.cpp
Comment (undertome) [inline src/backend/PostgresBackend.cpp:536]:
```suggestion
    std::uint32_t const ledgerSequence,
```

File: src/backend/PostgresBackend.cpp
Comment (undertome) [inline src/backend/PostgresBackend.cpp:713]:
```suggestion
    std::uint32_t const numLedgersToKeep,
```

File: src/backend/PostgresBackend.h
Comment (undertome) [inline src/backend/PostgresBackend.h:43]: Consider changing to `std::uint32_t const` wherever possible. I won't highlight every instance.

File: src/etl/ETLHelpers.h
Comment (undertome) [inline src/etl/ETLHelpers.h:200]: I'm in favor of replacing include guards with `#pragma once`. This is not necessary for this pull request, but we could note it as technical debt if everyone is on board with switching over. However, this does conflict with the convention in rippled. @cjcobb23 - thoughts?

File: src/rpc/RPCHelpers.cpp
Comment (undertome) [inline src/rpc/RPCHelpers.cpp:770]:
```suggestion
    bool const zeroIfFrozen,
```

File: src/rpc/RPCHelpers.cpp
Comment (undertome) [inline src/rpc/RPCHelpers.cpp:838]:
```suggestion
    std::uint32_t const ledgerSequence,
```

**Review by undertome — CHANGES_REQUESTED** [review]
Great job on this - I've submitted my initial review.

**Comment by undertome** [comment]
FWIW, try to avoid force-pushing while the pull request is being reviewed. This deletes history and can mangle pending reviews. The rippled approach is to include "fold" commits for changes that occur during the review: "[FOLD] parallelize db fetches". Once the pull request is approved you can squash the commits and force-push.

File: src/etl/ETLHelpers.h
Comment (cjcobb23) [reply src/etl/ETLHelpers.h:200]: Are there any downsides to `#pragma once`?

File: src/backend/BackendInterface.h
Comment (cjcobb23) [inline src/backend/BackendInterface.h:262]: Let's get rid of the yield context for writes. Only Postgres needs it, and thus Postgres can do that internally.

**Review by cjcobb23 — CHANGES_REQUESTED** [review]

File: src/backend/Pg.cpp
Comment (mtrippled) [reply src/backend/Pg.cpp:158]: Every new query does a connect() that ensures that there is a healthy connection (or throws). In this case, it's throwing so if the caller happens to retry with the same Pg object, then the connection status will again be checked and a reconnection attempt will be made.

File: src/backend/Pg.cpp
Comment (mtrippled) [reply src/backend/Pg.cpp:167]: see above

**Review by mtrippled — APPROVED** [review]
LGTM :+1:

File: src/backend/BackendInterface.h
Comment (natenichols) [reply src/backend/BackendInterface.h:237]: I'm gonna remove `yield_context` from all writes. They will be synchronous.

File: src/backend/PostgresBackend.cpp
Comment (natenichols) [reply src/backend/PostgresBackend.cpp:261]: I refactored this w/ the `synchronous`, and put it in `BackendInterface`
```cpp
    std::optional<LedgerRange>
    hardFetchLedgerRange() const
    {
        std::optional<LedgerRange> range = {};
        synchronous([&](boost::asio::yield_context yield)
        {
            range = hardFetchLedgerRange(yield);
        });

        return range;
    }
```

Which should call the underlying implementation for `hardFetchLedgerRange(yield_context)`

File: src/backend/CassandraBackend.h
Comment (natenichols) [reply src/backend/CassandraBackend.h:1125]: I actually just got rid of `syncRead`, it was only used twice.

File: src/backend/PostgresBackend.cpp
Comment (undertome) [inline src/backend/PostgresBackend.cpp:305]: Unresolving this request since it wasn't fixed.

File: src/backend/BackendInterface.h
Comment (undertome) [inline src/backend/BackendInterface.h]: Please change to std::uint32_t const for this instance and elsewhere in this file/PR.

File: src/backend/BackendInterface.h
Comment (undertome) [inline src/backend/BackendInterface.h:285]: Please change to `std::uint32_t const` for this instance and elsewhere in this file/PR.

**Review by undertome — CHANGES_REQUESTED** [review]
Thanks for addressing my changes. Just left a few comments. Almost there

File: src/etl/ETLHelpers.h
Comment (undertome) [reply src/etl/ETLHelpers.h:182]: I mean there are some issues, we can forget this for now. Resolving.

**Review by cjcobb23 — APPROVED** [review]
LGTM; great work!

**Pull request merged**
